modules/examples/sources/threaded-random-generator/threaded-random-generator.c

```diff
@@ -0,0 +1,185 @@
+#include "logthrsource/logthrsourcedrv.h"
+#include "atomic.h"
+#include "messages.h"
+#include "str-format.h"
+
+#include <strings.h>
+#include <sys/random.h>
+#include <unistd.h>
+
+struct ThreadedRandomGeneratorSourceDriver
+{
+  LogThreadedSourceDriver super;
+  GAtomicCounter exit_requested;
+
+  guint freq;
+  guint bytes;
+  gboolean urandom;
+};
+
+static gboolean
+_generate_random_bytes(guint8 *random, guint size, gboolean urandom)
+{
+  guint length = 0;
+
+  while (length < size)
+    {
+      gssize rc = getrandom(random + length, size - length, urandom ? 0 : GRND_RANDOM);
+
+      if (rc < 0)
+        {
+          msg_error("Could not generate random bytes", evt_tag_error("error"));
+          return FALSE;
+        }
+
+      length += rc;
+    }
+
+  return TRUE;
+}
+
+
+/* runs in a dedicated thread */
+static void
+_run(LogThreadedSourceDriver *s)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+
+  guint8 *random_bytes = g_malloc(self->bytes);
+
+  const gsize random_hex_str_size = self->bytes * 2 + 1;
+  gchar *random_hex_str = g_malloc(random_hex_str_size);
+
+  while (!g_atomic_counter_get(&self->exit_requested))
+    {
+      if (_generate_random_bytes(random_bytes, self->bytes, self->urandom))
+        {
+          format_hex_string(random_bytes, self->bytes, random_hex_str, random_hex_str_size);
+
+          LogMessage *msg = log_msg_new_empty();
+          log_msg_set_value(msg, LM_V_MESSAGE, random_hex_str, -1);
+
+          log_threaded_source_blocking_post(s, msg);
+        }
+
+      usleep(self->freq * 1000);
+    }
+
+  g_free(random_hex_str);
+  g_free(random_bytes);
+}
+
+static void
+_request_exit(LogThreadedSourceDriver *s)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+
+  g_atomic_counter_set(&self->exit_requested, TRUE);
+}
+
+static gboolean
+_init(LogPipe *s)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+
+  if (!self->bytes)
+    {
+      msg_error("The bytes() option for random-generator() is mandatory", log_pipe_location_tag(s));
+      return FALSE;
+    }
+
+  return log_threaded_source_driver_init_method(s);
+}
+
+static const gchar *
+_format_stats_instance(LogThreadedSourceDriver *s)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+  static gchar persist_name[1024];
+
+  if (s->super.super.super.persist_name)
+    g_snprintf(persist_name, sizeof(persist_name), "random-generator,%s", s->super.super.super.persist_name);
+  else
+    g_snprintf(persist_name, sizeof(persist_name), "random-generator,%u,%u,%d", self->bytes, self->freq, self->urandom);
+
+  return persist_name;
+}
+
+LogSourceOptions *
+threaded_random_generator_sd_get_source_options(LogDriver *s)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+
+  return &self->super.worker_options.super;
+}
```

Comment:
this should have been provided by LogThreadedSourceDriver, I think

Reply:
Right, I'll fix this.